core/src/sample_performance_service.rs

```diff
@@ -51,18 +51,19 @@ impl SamplePerformanceService {
         blockstore: &Arc<Blockstore>,
         exit: Arc<AtomicBool>,
     ) {
-        let (bank, highest_slot) = {
-            let forks = bank_forks.read().unwrap();
-            (forks.root_bank(), forks.highest_slot())
+        let root_bank = || -> Arc<Bank> {
+            let bank_forks = bank_forks.read().unwrap();
+            bank_forks.root_bank()
         };
+        let bank = root_bank();
```

Comment:
This is existing logic, but I notice it only now.
Would it unnecessarily hold a reference to the root bank that existing when the service started for the duration of the service execution?

In the loop below, `bank` is released at the end of the loop iteration.
But here, it is not.

I've created a PR to refactor this code a bit more, hopefully, to make it easier to read: https://github.com/solana-labs/solana/pull/30297

Reply:
> Would it unnecessarily hold a reference to the root bank that existing when the service started for the duration of the service execution?

I think you're right, good catch. That behavior looks to have been in there since the creation of this service. And to confirm, this service runs as long as the process, so that ref is never being released. One last note to recall is that this service is only started on RPC nodes, not regular validators.

In addition to some memory that is never given back, I believe this `Bank` being alive will be preventing some `AccountsDb` cleanup from occurring (ie cleanup of old AppendVecs). We can probably ask AccountsDb folks for a better assessment of the potential damage in that area (who knows, this might resolve some ghosts they've been chasing), but I think we may consider a very simple backport to v1.14 and v1.15. Maybe just a 1-liner like `drop(bank);` before we enter the loop.

Reply:
> > Would it unnecessarily hold a reference to the root bank that existing when the service started for the duration of the service execution?
> 
> I think you're right, good catch. That behavior looks to have been in there since the creation of this service. And to confirm, this service runs as long as the process, so that ref is never being released. One last note to recall is that this service is only started on RPC nodes, not regular validators.

This is interesting.  Thank you for checking and sharing.

> In addition to some memory that is never given back, I believe this `Bank` being alive will be preventing some `AccountsDb` cleanup from occurring (ie cleanup of old AppendVecs). We can probably ask AccountsDb folks for a better assessment of the potential damage in that area (who knows, this might resolve some ghosts they've been chasing), but I think we may consider a very simple backport to v1.14 and v1.15. Maybe just a 1-liner like `drop(bank);` before we enter the loop.

You or Tyera will have to make a call here :)
In my personal Rust style-guide book, `drop()` is next to `unsafe` :))
Though, from the practical standpoint, it could be the simplest thing to backport, yes.

I have experience maintaining multiple branches of a big product over multiple years and cherry-picking changes.
From this experience, a cherry-pick that is syntactically different is the number one thing that I would avoid at all costs.
So, if I would be the person maintaining older versions, I would put as much effort as is reasonable, to cherry-pick only full changes.  Preserving their syntactic form as much as possible.
Otherwise, it is really easy to end up in a merge hell.
My experience is that any non-trivial backporting is already very hard, even when changes can be cherry-picked with no to minimal modifications.

Reply:
Let's not necropost on this closed PR. Comments here, please: https://github.com/solana-labs/solana/issues/30312